crates/uv-resolver/src/pubgrub/dependencies.rs

```diff
@@ -30,14 +30,28 @@ impl PubGrubDependency {
         dev: Option<&'a GroupName>,
         source_name: Option<&'a PackageName>,
     ) -> impl Iterator<Item = Self> + 'a {
-        let iter = if requirement.extras.is_empty() {
-            Either::Left(iter::once(None))
+        let iter = if !requirement.extras.is_empty() {
+            Either::Left(Either::Left(
+                requirement
+                    .extras
+                    .clone()
+                    .into_iter()
+                    .map(|extra| (Some(extra), None)),
```

Comment:
If `requirement.groups` isn't empty at this point, does this mean that all of those groups get skipped?

Reply:
Interesting. Hard to tell from the type definition. But encoding that invariant in the types is probably annoying.

Reply:
I could do it? It's a bit tedious because these also get serialized so we'd want to maintain backwards compatibility, but I could do it. Somewhat prefer to just document it clearly though.

Reply:
Documenting it seems fine, yeah.

I am pretty much always torn about how much to encode in the types. I do like it, because I find it makes refactoring without fear a lot easier. It prevents you from getting the case analysis wrong. But as you rightly point out, it's a pain. And if the invariants change, it can make refactoring a pain (albeit higher confidence) too.

It's a constant balancing act.

Reply:
Added.